fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java

```diff
@@ -146,59 +147,46 @@ public ScalarOperator reduceDateToDatetimeCast(BinaryPredicateOperator operator)
-                resultBinaryPredicateOperator =
-                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.GE, castChild, newDate);
+                newDate = ConstantOperator.createDate(targetDateTime);
+                resultBinaryPredicateOperator = BinaryPredicateOperator.ge(castChild, newDate);
                 break;
             case LE:
                 // when the BinaryType is <= ,cast dateTime to maximum date type；
                 // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
                 // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-27)
                 offset = 0;
                 targetDateTime = bottomDateTime.plusDays(offset);
-                newDate = ConstantOperator.createDate(targetDateTime.truncatedTo(ChronoUnit.DAYS));
-                resultBinaryPredicateOperator =
-                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
+                newDate = ConstantOperator.createDate(targetDateTime);
+                resultBinaryPredicateOperator = BinaryPredicateOperator.le(castChild, newDate);
                 break;
             case LT:
                 // when the BinaryType is < ,cast dateTime to maximum date type；
                 // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-27)
                 // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
-                if (originalDateTime.isEqual(bottomDateTime)) {
-                    offset = -1;
-                } else {
-                    offset = 0;
-                }
+                offset = originalDateTime.isEqual(bottomDateTime) ? -1 : 0;
                 targetDateTime = bottomDateTime.plusDays(offset);
-                newDate = ConstantOperator.createDate(targetDateTime.truncatedTo(ChronoUnit.DAYS));
-                resultBinaryPredicateOperator =
-                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
+                newDate = ConstantOperator.createDate(targetDateTime);
+                resultBinaryPredicateOperator = BinaryPredicateOperator.le(castChild, newDate);
                 break;
             case EQ:
                 // when the BinaryType is = ,cast dateTime to equivalent date type；
                 // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
                 if (!originalDateTime.isEqual(bottomDateTime)) {
                     resultBinaryPredicateOperator = operator;
                 } else {
-                    newDate = ConstantOperator.createDate(bottomDateTime.truncatedTo(ChronoUnit.DAYS));
-                    resultBinaryPredicateOperator =
-                            new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.EQ, castChild, newDate);
+                    newDate = ConstantOperator.createDate(bottomDateTime);
```

Comment:
Why bottomDateTime.truncatedTo(ChronoUnit.DAYS) was originally needed

Reply:
I think you should read the comment about the rule